unified_planning/model/htn/method.py

```diff
@@ -0,0 +1,239 @@
+    """This is the method interface."""
+    def __init__(self, _name: str, _parameters: 'OrderedDict[str, up.model.types.Type]' = None,
+                 _env: Environment = None, **kwargs: 'up.model.types.Type'):
+        self._env = get_env(_env)
+        self._task: Optional[ParameterizedTask] = None
+        self._name = _name
+        self._parameters: 'OrderedDict[str, Parameter]' = OrderedDict()
+        self._preconditions: List[up.model.fnode.FNode] = []
+        self._constraints: List[up.model.fnode.FNode] = []
+        self._subtasks: List[Subtask] = []
+        if _parameters is not None:
+            assert len(kwargs) == 0
+            for n, t in _parameters.items():
+                self._parameters[n] = Parameter(n, t)
+        else:
+            for n, t in kwargs.items():
+                self._parameters[n] = Parameter(n, t)
+
+    def __repr__(self) -> str:
+        s = []
+        s.append(f'method {self.name}')
+        first = True
+        for p in self.parameters:
+            if first:
+                s.append('(')
+                first = False
+            else:
+                s.append(', ')
+            s.append(str(p))
+        if not first:
+            s.append(')')
+        s.append(' {\n')
+        s.append(f'  task = {self._task}\n')
+        s.append('  preconditions = [\n')
+        for c in self.preconditions:
+            s.append(f'    {str(c)}\n')
+        s.append('  ]\n')
+        s.append('  constraints = [\n')
+        for c in self.constraints:
+            s.append(f'    {str(c)}\n')
+        s.append('  ]\n')
+        s.append('  subtasks = [\n')
+        for st in self.subtasks:
+            s.append(f'      {str(st)}\n')
+        s.append('  ]\n')
+        s.append('}')
+        return ''.join(s)
+
+    def __eq__(self, oth: object) -> bool:
+        if not isinstance(oth, Method):
+            return False
+        return (self._env == oth._env and
+                self._name == oth._name and
+                self._parameters == oth._parameters and
+                self._task == oth._task and
+                set(self._preconditions) == set(oth._preconditions) and
+                set(self._subtasks) == set(oth._subtasks) and
+                set(self._constraints) == set(oth._constraints))
+
+    def __hash__(self) -> int:
+        res = hash(self._name)
+        res += hash(self._task)
+        res += sum(map(hash, self.parameters))
+        res += sum(map(hash, self._preconditions))
+        res += sum(map(hash, self._constraints))
+        res += sum(map(hash, self.subtasks))
+        return res
+
+    @property
+    def name(self) -> str:
+        """Returns the action name."""
+        return self._name
+
+    @property
+    def achieved_task(self) -> ParameterizedTask:
+        """Returns the task that this method achieves."""
+        assert self._task is not None
+        return self._task
+
+    def set_task(self, task: Union[Task, ParameterizedTask], *arguments: Parameter):
```

Comment:
I added a TODO referencing issue #153 that would describe the missing feature.

Reply:
The TODO is OK for now then. (Having the environment in FNode is indeed a good idea, we "copied" expressions from pySMT, and there there is no env inside because of efficiency reasons, but in planning expressions are generally way samller than in SMT )